File: app/src/main/java/com/chetaru/tribe365_new/API/retrofit/ApiClient.java
package com.chetaru.tribe365_new.API.retrofit;

import java.util.concurrent.TimeUnit;

import okhttp3.OkHttpClient;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class ApiClient {


    public static final String BASE_URL = "https://console.tribe365.co/api/";  // live URL

    //public static final String BASE_URL = "https://tribe365demo.chetaru.co.uk/api/"; //(demo)  staging

    //public static final String BASE_URL = "https://staging-redis-console.tribe365.co/api/"; //new staging Redis

    //only use for help
    public static final String SUPPORT_URL = "https://console.tribe365.co/resources/views/support/";  //support url Call for help

    public static OkHttpClient okHttpClient = null;
    private static Retrofit retrofit = null;


    public static Retrofit getClient() {
        if (retrofit == null) {
            retrofit = new Retrofit.Builder()
                    .baseUrl(BASE_URL)
                    .client(getRequestHeader())
                    .addConverterFactory(GsonConverterFactory.create())
                    .build();
        }
        return retrofit;
    }


    public static Retrofit getCustomClient(String baseURL) {
        Retrofit retrofit_ = new Retrofit.Builder()
                .baseUrl(baseURL)
                .client(getRequestHeader())
                .addConverterFactory(GsonConverterFactory.create())
                .build();
        return retrofit_;
    }

    public static OkHttpClient getRequestHeader() {
        if (null == okHttpClient) {
            okHttpClient = new OkHttpClient.Builder()
                    .readTimeout(60, TimeUnit.SECONDS)
                    .connectTimeout(60, TimeUnit.SECONDS)
                    .build();
        }
        return okHttpClient;
    }
}

File: app/src/main/java/com/chetaru/tribe365_new/KotlinCode/APi/ApiClient.kt
package com.chetaru.tribe365_new.KotlinCode.APi

import android.annotation.SuppressLint
import android.content.Context
import com.chetaru.tribe365_new.KotlinCode.ApiCall.API
import okhttp3.Cache
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.util.concurrent.TimeUnit

@SuppressLint("StaticFieldLeak")
object ApiClient {
    private var context: Context?= null

     fun init(context: Context){ApiClient.context = context}

    private fun retrofitService():Retrofit{
        return Retrofit.Builder().addConverterFactory(GsonConverterFactory.create())
            .client(okHttpClient(context!!))
            .baseUrl(BASE_URL)
            .build()
    }
    private fun okHttpClient(context: Context): OkHttpClient{
        val cashSize: Long= (5 * 1024 * 1024).toLong()
        val cache = Cache(context.cacheDir,cashSize)
        return OkHttpClient.Builder()
            .cache(cache)
            .addInterceptor{ chain: Interceptor.Chain ->
                var request= chain.request()
                request = if (CheckNetwork.isConnected(context)){
                    request.newBuilder()
                        .header("content-type","application-json")
                        .header("Cache-control","public, max-age=" + 5)
                        .build()
                }else{
                    request.newBuilder()
                        .header("content-type","application-json")
                        .header("Cache-control","public, only-if-cached, max-stale= " + 60*60*24*7)
                        .build()
                }
                chain.proceed(request)
            }.readTimeout(30,TimeUnit.SECONDS)
            .connectTimeout(10,TimeUnit.SECONDS)
            .build()
    }
    val instance: API by lazy {
        retrofitService().create(API::class.java)
    }
}